Zendesk Mobile: register devices for push notifications

ScoutHarris opened the pull request:
Fixes #9242 

To test:
- Launch the app. 
- Login.
- Agree to notifications if necessary.
- Verify you see this message in the logs:
`Zendesk successfully registered device:<device_id>`

I did add a method to un-register devices. At this time it is not called. However, I believe it will be used in the future, so I'm adding it now. If you feel spunky and want to test it:
- Add this code to `PushNotificationsManager:unregisterDeviceToken`
```
        if let deviceToken = deviceToken {
            ZendeskUtils.unregisterDevice(deviceToken)
        }
```
- Logout.
- Notice the log message `Zendesk successfully unregistered device:<device_id>`

frosty commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -85,6 +85,34 @@ import CoreTelephony
 
+    static func registerDevice(_ identifier: String) {
+        if ZDKConfig.instance().userIdentity == nil {
+            ZendeskUtils.createZendeskIdentity(withUser: false)
```
Is this ever likely to happen, that we'll have no identity? Is there a reason we're creating an anonymous identity here instead of using `createIdentity` in `ZendeskUtils` to create one based on the user's site / accounts?

ScoutHarris replied:
Hey @frosty . When registering a device with ZD, the user information is irrelevant. ZD just needs the device ID and an instance of `ZDKConfig.instance().userIdentity`. It did happen while I was testing it. 

However, I just learned today that WP PNs require a .com account. So this may become moot anyway. I'll be making changes in light if my recent education.

ScoutHarris commented:
Hey @frosty . I made the following changes (and updated the description accordingly):
- Only register with Zendesk if there is a .com account.
- Unregister with Zendesk if the user logs out of .com account.

Referring back to our discussion about creating the identity when (un)registering, it turns out ZD doesn't give a hoot if `ZDKConfig.instance().userIdentity` is set or not in these cases. In addition, it gets persnickity when it is changed. So I removed that altogether from the register methods.

I'd appreciate another review. Thank you!

frosty commented:
I just tested this again, and it looks like we _do_ need to set an identity – I got this output in the console when registering:

```
2018-05-03 11:16:54:264 WordPress[11796:3426059] Zendesk couldn't register device: 15e3b7c2ac1b98e8b8d5aa63afb73dc42f83f9d927410b4ea9bf7c58f4bf5214. Error: Error Domain=ZDDispatcher Code=-1013 "(null)" UserInfo={zd_dispatcher_err=Anonymous identity is nil. Have you set an identity?}
```

ScoutHarris commented:
> Zendesk couldn't register device

Darn it. I never once saw that! I'll take another look. Thanks @frosty !

ScoutHarris commented:
Hey @frosty I just realized you _may_ come across this - Zendesk does not allow ticket creation with Agent accounts, so you'll have to use an email address that is not associated with your Zendesk agent account (if you even have one).

frosty commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -13,6 +13,8 @@ import CoreTelephony
 
     private var userName: String?
     private var userEmail: String?
+    private var needToRegisterDevice = false
```
Do we need both properties here? Could we just check whether we have a deviceID, and if so use it to register the device?

ScoutHarris replied:
Oh. Hehe. Good point.

frosty commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -69,20 +71,35 @@ import CoreTelephony
+
+    static func setNeedToRegisterDevice(_ identifier: String) {
+        ZendeskUtils.sharedInstance.needToRegisterDevice = true
+        ZendeskUtils.sharedInstance.deviceID = identifier
```
Should we call `createZendeskIdentity` here so we can go ahead and create one if we have everything we need at this point?

ScoutHarris replied:
Well, the point of postponing device registration (and thus identity creation) is that at this point the user's account isn't fully loaded, so the identity won't be created anyway. When a user accesses a Support option, the identity is then created and the device registered if that's successful. Doing them in tandem works best. I don't see a reason to do it otherwise...

frosty replied:
If this isn't the first launch of the app (so the user is already logged in), then we'll have account details ready?

ScoutHarris replied:
Sure, probably. Maybe I'm not understanding, but I don't see what it gains us.  `createZendeskIdentity` is called when a ZD view is needed. 

Plus, now that I think about it, creating the identity this early may cause issues with self-hosted users. In this case, we save the user information used for Zendesk to User Defaults. If we create the identity, they change or add sites, then access Support, we've potentially saved the wrong information to UD, thus using the wrong information for Zendesk.

frosty replied:
Okay, sounds good – just wanted to check if we could do it sooner :) Totally makes sense!

ScoutHarris commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -191,16 +189,14 @@ private extension ZendeskUtils {
             if let error = error {
                 DDLogInfo("Zendesk couldn't register device: \(deviceID). Error: \(error)")
             } else {
-                ZendeskUtils.sharedInstance.needToRegisterDevice = false
```
Probably should. Zendesk just ignores it if it's already registered but no sense in making an extra call. Will do!
